pkg/pebinaryexecutor/pebinary.go

```diff
@@ -212,22 +223,110 @@ func (p *PEBinaryConfig) KubeProxy(ctx context.Context, args []string) error {
 
 	args = append(getArgs(extraArgs), args...)
 
-	logrus.Infof("Running RKE2 kube-proxy %s", args)
-	go func() {
+	win.ProcessWaitGroup.StartWithContext(ctx, func(ctx context.Context) {
 		outputFile := logging.GetLogger(filepath.Join(p.DataDir, "agent", "logs", "kube-proxy.log"), 50)
 		for {
+			logrus.Infof("Running RKE2 kube-proxy %s", args)
 			cmd := exec.CommandContext(ctx, filepath.Join("c:\\", p.DataDir, "bin", "kube-proxy.exe"), args...)
 			cmd.Stdout = outputFile
 			cmd.Stderr = outputFile
 			err := cmd.Run()
 			logrus.Errorf("kube-proxy exited: %v", err)
-			time.Sleep(5 * time.Second)
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				time.Sleep(5 * time.Second)
+			}
 		}
-	}()
+	})
 
 	return nil
 }
 
+// Docker starts cri-dockerd as implemented in the k3s cridockerd package
+func (p *PEBinaryConfig) Docker(ctx context.Context, cfg *config.Node) error {
+	return cridockerd.Run(ctx, cfg)
+}
+
+// Containerd configures and starts containerd.
+func (p *PEBinaryConfig) Containerd(ctx context.Context, cfg *config.Node) error {
+	args := getContainerdArgs(cfg)
+	stdOut := io.Writer(os.Stdout)
+	stdErr := io.Writer(os.Stderr)
+
+	if cfg.Containerd.Log != "" {
+		logrus.Infof("Logging containerd to %s", cfg.Containerd.Log)
+		fileOut := &lumberjack.Logger{
+			Filename:   cfg.Containerd.Log,
+			MaxSize:    50,
+			MaxBackups: 3,
+			MaxAge:     28,
+			Compress:   true,
+		}
+
+		// If rke2 is started with --debug, write logs to both the log file and stdout/stderr,
+		// even if a log path is set.
+		if cfg.Containerd.Debug {
+			stdOut = io.MultiWriter(stdOut, fileOut)
+			stdErr = io.MultiWriter(stdErr, fileOut)
+		} else {
+			stdOut = fileOut
+			stdErr = fileOut
+		}
+	}
+
+	win.ProcessWaitGroup.StartWithContext(ctx, func(ctx context.Context) {
+		env := []string{}
+		cenv := []string{}
+
+		for _, e := range os.Environ() {
+			pair := strings.SplitN(e, "=", 2)
+			switch {
+			case pair[0] == "NOTIFY_SOCKET":
+				// elide NOTIFY_SOCKET to prevent spurious notifications to systemd
+			case pair[0] == "CONTAINERD_LOG_LEVEL":
+				// Turn CONTAINERD_LOG_LEVEL variable into log-level flag
+				args = append(args, "--log-level", pair[1])
+			case strings.HasPrefix(pair[0], "CONTAINERD_"):
+				// Strip variables with CONTAINERD_ prefix before passing through
+				// This allows doing things like setting a proxy for image pulls by setting
+				// CONTAINERD_https_proxy=http://proxy.example.com:8080
+				pair[0] = strings.TrimPrefix(pair[0], "CONTAINERD_")
+				cenv = append(cenv, strings.Join(pair, "="))
+			default:
+				env = append(env, strings.Join(pair, "="))
```

Comment:
shouldn't it only check on `CONTAINERD_` envs?

Reply:
This is a mirror implementation of what k3s does, I'm not sure what the implications are if start ignoring specific environment variables